system/trace/trace_dump.c

```diff
@@ -0,0 +1,570 @@
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <nuttx/sched_note.h>
+
+#include "trace.h"
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+#  ifdef CONFIG_LIB_SYSCALL
+#    include <syscall.h>
+#  else
+#    define CONFIG_LIB_SYSCALL
+#    include <syscall.h>
+#    undef CONFIG_LIB_SYSCALL
+#  endif
+#endif
+
+#ifdef CONFIG_SMP
+#  define NCPUS CONFIG_SMP_NCPUS
+#else
+#  define NCPUS 1
+#endif
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Renumber idle task PIDs
+ *  In NuttX, PID number less than NCPUS are idle tasks.
+ *  In Linux, there is only one idle task of PID 0.
+ */
+
+#define get_pid(pid)  ((pid) < NCPUS ? 0 : (pid))
+
+#define get_task_state(s) ((s) <= LAST_READY_TO_RUN_STATE ? 'R' : 'S')
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* The structure to hold the context data of trace dump */
+
+struct trace_dump_cpu_context_s
+{
+  bool ininterrupt;       /* In interrupt handler flag */
+  bool pendingswitch;     /* sched_switch pending flag */
+  int current_state;      /* Task state of the current line */
+  pid_t current_pid;      /* Task PID of the current line */
+  pid_t next_pid;         /* Task PID of the next line */
+};
+
+struct trace_dump_task_context_s
+{
+  FAR struct trace_dump_task_context_s *next;
+  pid_t pid;                              /* Task PID */
+  int syscall_nest;                       /* Syscall nest level */
+  char name[CONFIG_TASK_NAME_SIZE + 1];   /* Task name (with NUL terminator) */
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct trace_dump_cpu_context_s g_dump_cpu_ctx[NCPUS];
+static FAR struct trace_dump_task_context_s *g_dump_task_ctx = NULL;
```

Comment:
let's change to local variable?

Reply:
You mean both g_dump_cpu_ctx[] and g_dump_task_ctx?
It seems a bit complicated to add both argument to every private functions.
I think we have no strong reason to change into local variable because:
- These are widely accessed by private functions in this file, 
- Already defined as static and used only in this file.
- The function never called recursively.